src/Npgsql/TypeHandlers/RangeHandler.cs

```diff
@@ -181,4 +154,35 @@ public async Task Write(NpgsqlRange<TElement> value, NpgsqlWriteBuffer buf, Npgs
 
         #endregion
     }
+
+    interface IRangeHandlerWithTwoElements
```

Comment:
Looks ugly.

Reply:
I kind of agree.
The ugliness lies in the fact that - as I understand it - it doesn't really open up the handler to support any base type for ranges.

What if I opened an issue to support NpgsqlRange\<LocalDateTime> in the NodaTime plugin?

Would we then make it a IRangeHandlerWithThreeElements or would we just write an new handler for the plugin and hook that one up instead of the normal one?

Reply:
Why not change the interface to following (the `fieldDescription` is superfluous in this interface anyways):

```csharp
public interface IRangeHandlerWithMultipleElements
{
    IEnumerable<Type> AdditionalFieldTypes { get; }
}
```

Then we could do range handlers that support as many types as we deem necessary like the following:
```csharp
public class RangeHandler<TElement1, TElement2> : RangeHandler<TElement1>, INpgsqlTypeHandler<NpgsqlRange<TElement2>>,
    IRangeHandlerWithMultipleElements
{
    public virtual IEnumerable<Type> AdditionalFieldTypes
    {
        get { yield return typeof(NpgsqlRange<TElement2>); }
    }

    public RangeHandler(NpgsqlTypeHandler elementHandler) : base(elementHandler) {}

    ValueTask<NpgsqlRange<TElement2>> INpgsqlTypeHandler<NpgsqlRange<TElement2>>.Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription)
        => DoRead<TElement2>(buf, len, async, fieldDescription);

    public int ValidateAndGetLength(NpgsqlRange<TElement2> value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
        => ValidateAndGetLength<TElement2>(value, ref lengthCache, parameter);

    public Task Write(NpgsqlRange<TElement2> value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
        => Write<TElement2>(value, buf, lengthCache, parameter, async);
}

public class RangeHandler<TElement1, TElement2, TElement3> : RangeHandler<TElement1, TElement2>, INpgsqlTypeHandler<NpgsqlRange<TElement3>>,
    IRangeHandlerWithMultipleElements
{
    public override IEnumerable<Type> AdditionalFieldTypes
    {
        get
        {
            yield return typeof(NpgsqlRange<TElement2>);
            yield return typeof(NpgsqlRange<TElement3>);
        }
    }

    public RangeHandler(NpgsqlTypeHandler elementHandler) : base(elementHandler) {}

    ValueTask<NpgsqlRange<TElement3>> INpgsqlTypeHandler<NpgsqlRange<TElement3>>.Read(NpgsqlReadBuffer buf, int len, bool async, FieldDescription fieldDescription)
        => DoRead<TElement3>(buf, len, async, fieldDescription);

    public int ValidateAndGetLength(NpgsqlRange<TElement3> value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
        => ValidateAndGetLength<TElement3>(value, ref lengthCache, parameter);

    public Task Write(NpgsqlRange<TElement3> value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
        => Write<TElement3>(value, buf, lengthCache, parameter, async);
}
```

Reply:
You guys are right, I did this a bit quickly without thinking more generally.

I've just pushed a commit that follows much of what @Brar proposed - take a look. Since our type handler system is generics-based, if we want a range handler for 3 types we need a new RangeHandler class (much like there are lots of ValueTuple classes with different numbers of type parameters). I don't think there's a way around that.

I also fixed an important problem with the previous implementation - we were binding the second range CLR type without any sort of check in element's handler's bound CLR types: that's incorrect. The idea is that the handlers themselves don't decide which CLR types they bind to - the user has control over that. So we now filter the range handler's exposed list via its element handler's list, and only bind range types whose corresponding element types have been bound. Hope that's clear, it's definitely a very elaborate type handling model :)

The fieldDescription is basically there for BitStringHandler, which returns a different type based on the type modifier (a single bool for one, or a BitArray otherwise). So in theory we don't support a range that will automatically be over bool or BitArray based on the fieldDescription - I think that's OK :)

Reply:
> Since our type handler system is generics-based, if we want a range handler for 3 types we need a new RangeHandler class (much like there are lots of ValueTuple classes with different numbers of type parameters). I don't think there's a way around that.

That's a pretty common scenario. Also see `Action`, `Func` and others.
I don't think that we will bind like hundreds of types to `RangeHandler` so that seems absolutely ok to me.

> So we now filter the range handler's exposed list via its element handler's list, and only bind range types whose corresponding element types have been bound.

Makes sense.

> So in theory we don't support a range that will automatically be over bool or BitArray based on the fieldDescription - I think that's OK :)

I wouldn't know what to do with such a beast 😜